### PR TITLE
Multiple notification channels

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
## Description

Requires https://github.com/ibi-group/otp-middleware/pull/185.

This PR adds push notifications as an option, describing the number of push devices registered, and allows setting multiple notification channels.

## PR Checklist
- [x] Does the code follow accessibility standards (WCAG 2.1 AA Compliant)? => `aria-describedby` used with labels for each notification channel.
- [x] Are all languages supported (Internationalization/Localization)? => i18n supported but not all languages are filled.
- [x] Are appropriate Typescript types implemented?

| Before | After |
|--------|-------|
|  Before  | After |

### CODE REVIEW

File: lib/components/user/notification-prefs-pane.tsx [hunk 1]
```diff
@@ -1,119 +1,132 @@
...
-  }
   onRequestPhoneVerificationCode: PhoneCodeRequestHandler
   onSendPhoneVerificationCode: PhoneVerificationSubmitHandler
   phoneFormatOptions: {
     countryCode: string
   }
 }
 
-const allowedNotificationChannels = ['email', 'sms', 'none']
+const allowedNotificationChannels = ['email', 'sms', 'push']
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/notification-prefs-pane.tsx:30]: If the value is `'none'` as in the current implementation (or no known value is present), no options will be checked. Now, if no options are checked, should we send `''` or `'none'` to the server?

[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/notification-prefs-pane.tsx:30]: Looking at https://github.com/ibi-group/otp-middleware/pull/185, it seems like either way is possible, however I would side with sending an empty string.

File: lib/components/user/notification-prefs-pane.tsx [hunk 2]
```diff
@@ -22,91 +21,109 @@ interface Props extends FormikProps<User> {
...
+                value={type}
+              />
+            </span>
+            <span>
+              {type === 'email' ? (
+                <>
+                  <label htmlFor={inputId}>
                     <FormattedMessage id="common.notifications.email" />
-                  ) : type === 'sms' ? (
+                  </label>
```
Comment (amy-corson-ibigroup) [inline lib/components/user/notification-prefs-pane.tsx:90]: Might be a bit cleaner to move the `label` out of these blocks, and conditionally render the i18n id

File: lib/components/user/notification-prefs-pane.tsx [hunk 3]
```diff
@@ -22,91 +21,109 @@ interface Props extends FormikProps<User> {
...
+      <legend>
+        <FormattedMessage id="components.NotificationPrefsPane.notificationChannelPrompt" />
+      </legend>
+      {allowedNotificationChannels.map((type) => {
+        // TODO: If removing the Save/Cancel buttons on the account screen,
+        // persist changes immediately when onChange is triggered.
+        const inputId = `notification-channel-${type}`
+        const inputDescriptionId = `${inputId}-description`
+        return (
+          <NotificationOption key={type}>
```
Comment (amy-corson-ibigroup) [inline lib/components/user/notification-prefs-pane.tsx:73]: Should these options be in a `<ul>`?

**Review by amy-corson-ibigroup — APPROVED** [review]
A couple of small things but largely I think it looks good and seems to work well, pending push notification support from the middleware and mobile apps!

File: lib/components/user/notification-prefs-pane.tsx (hunk 2 above)
Comment (binh-dam-ibigroup) [reply lib/components/user/notification-prefs-pane.tsx]: Good idea! Updated in https://github.com/opentripplanner/otp-react-redux/pull/899/commits/9ae5bfda7a0b9fff321465c130bb0810da4a105c. It also unveiled an oversight in `TripNotificationsPane` that I fixe with https://github.com/opentripplanner/otp-react-redux/pull/899/commits/bf54d586f1847e3bbe1b012118096ef98f0edb00.

**Review by daniel-heppner-ibigroup — CHANGES_REQUESTED** [review]
My big question is here is about the push notifications showing up for OTP-RR deployments with no mobile app. I'm worried about the implication of a mobile app existing if a user sees "Register your device using the mobile app" when it doesn't exist for them. 

Related to design, I think there is too much vertical space around the phone number options, which makes it easy to miss the push notifications checkbox entirely. I think that's especially the case in your screenshot with more stuff going on. (GitHub won't let me upload a screenshot... will try again later)

File: i18n/i18n-exceptions.json [hunk 4]
```diff
@@ -1,5 +1,10 @@
 {
   "groups": {
+    "common.notifications.*": [
+      "email",
+      "sms",
+      "push"
+    ],
```
Comment (daniel-heppner-ibigroup) [inline i18n/i18n-exceptions.json:7]: Is this still needed?

**Review by daniel-heppner-ibigroup — CHANGES_REQUESTED** [review]

File: lib/components/user/user-account-screen.js [hunk 5]
```diff
@@ -43,9 +43,17 @@ const validationSchema = yup.object({
...
-    // TODO: Change state of Save button while the update action takes place.
 
-    await createOrUpdateUser(userData, silentOnSucceed, intl)
+    // Convert the notification attributes from array to comma-separated string.
+    const passedUserData = clone(userData)
+    const { notificationChannel } = passedUserData
+    if (
+      typeof notificationChannel === 'object' &&
+      typeof notificationChannel.length === 'number'
+    ) {
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/user-account-screen.js:53]: The selected channels are persisted by OTP-middleware as a comma-separated string, see https://github.com/ibi-group/otp-middleware/blob/be241ed8e284281eb6420adb419a7a030c23d297/src/main/java/org/opentripplanner/middleware/models/OtpUser.java#L43, so compatibility with existing accounts is maintained.

[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/user-account-screen.js:53]: OTP-middleware handles the conversion so it returns an array to the UI.

File: lib/components/user/notification-prefs-pane.tsx [hunk 6]
```diff
@@ -22,91 +21,97 @@ interface Props extends FormikProps<User> {
   }
 }
 
-const allowedNotificationChannels = ['email', 'sms', 'none']
+const allowedNotificationChannels = ['email', 'sms', 'push']
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/components/user/notification-prefs-pane.tsx]: Very good point. I have added a config parameter in https://github.com/opentripplanner/otp-react-redux/pull/899/commits/6954113e49e8541bdf35af84cd6c24b4a0d48f68.

**Comment by binh-dam-ibigroup** [comment]
> Related to design, I think there is too much vertical space around the phone number options, which makes it easy to miss the push notifications checkbox entirely.

@daniel-heppner-ibigroup I tried adding visual grouping in https://github.com/opentripplanner/otp-react-redux/pull/899/commits/344442f42157939bef5dd4fc69bfcbdb6fe36ff2. What do you think?

**Pull request merged**